frontend/exporter/src/id_table.rs

```diff
@@ -0,0 +1,375 @@
+/// inside it. Serializing `WithTable<T>` will serialize IDs only,
+/// skipping values. Deserialization of a `WithTable<T>` will
+/// automatically use the table and IDs to reconstruct skipped values.
+#[derive(Serialize, Debug)]
+pub struct WithTable<T> {
+    table: Table,
+    value: T,
+}
+
+/// The state used for deserialization: a table.
+static DESERIALIZATION_STATE: LazyLock<Mutex<Table>> =
+    LazyLock::new(|| Mutex::new(Table::default()));
+static DESERIALIZATION_STATE_LOCK: LazyLock<Mutex<()>> = LazyLock::new(|| Mutex::new(()));
+
+/// The mode of serialization: should `Node<T>` ship values of type `T` or not?
+static SERIALIZATION_MODE_USE_IDS: std::sync::atomic::AtomicBool =
+    std::sync::atomic::AtomicBool::new(false);
+
+fn serialize_use_id() -> bool {
+    SERIALIZATION_MODE_USE_IDS.load(Ordering::Relaxed)
+}
+
+impl<T> WithTable<T> {
+    /// Runs `f` with a `WithTable<T>` created out of `map` and
+    /// `value`. Any serialization of values of type `Node<_>` will
+    /// skip the field `value`.
+    pub fn run<R>(map: Table, value: T, f: impl FnOnce(&Self) -> R) -> R {
+        if serialize_use_id() {
+            panic!("CACHE_MAP_LOCK: only one WithTable serialization can occur at a time (nesting is forbidden)")
+        }
+        SERIALIZATION_MODE_USE_IDS.store(true, Ordering::Relaxed);
+        let result = f(&Self { table: map, value });
+        SERIALIZATION_MODE_USE_IDS.store(false, Ordering::Relaxed);
+        result
+    }
+    pub fn destruct(self) -> (T, Table) {
+        let Self { value, table: map } = self;
+        (value, map)
+    }
+}
+
+/// Helper function that makes sure no nested deserializations occur.
+fn full_id_deserialization<T>(f: impl FnOnce() -> T) -> T {
+    let _lock = *DESERIALIZATION_STATE_LOCK.lock().expect("CACHE_MAP_LOCK: only one WithTable deserialization can occur at a time (nesting is forbidden)");
+    let result = f();
+    result
+}
+
+/// The deserializer of `WithTable<T>` is special. First, we need to continuously
+impl<'de, T: Deserialize<'de>> serde::Deserialize<'de> for WithTable<T> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        use serde::de::{self, Deserialize, MapAccess, SeqAccess, Visitor};
+        use std::fmt;
+        use std::marker::PhantomData;
+        #[derive(Deserialize, Debug)]
+        #[serde(field_identifier, rename_all = "lowercase")]
+        enum Field {
+            Table,
+            Value,
+        }
+
+        struct WithTableVisitor<T>(PhantomData<T>);
+
+        impl<'de, T: Deserialize<'de>> Visitor<'de> for WithTableVisitor<T> {
+            type Value = WithTable<T>;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("struct WithTable")
+            }
+
+            fn visit_seq<V>(self, mut seq: V) -> Result<Self::Value, V::Error>
```

Comment:
why do you need both `visit_seq` and `visit_map`? shouldn't one of them be enough?

Reply:
I see!